src/persist-client/src/batch.rs

```diff
@@ -898,16 +898,58 @@ impl<T: Timestamp + Codec64> BatchParts<T> {
         let index = u64::cast_from(self.finished_parts.len() + self.writing_parts.len());
         let ts_rewrite = None;
 
-        let handle = if updates.goodbytes() < self.cfg.inline_writes_single_max_bytes {
+        let should_columnar_encode = self.cfg.batch_columnar_format.is_structured()
+            && !self.cfg.batch_columnar_stats_only_override;
+
+        // If we're going to encode structured data then halve our limit since we're storing
+        // it twice, once as binary encoded and once as structured.
+        let inline_threshold = if should_columnar_encode {
+            self.cfg.inline_writes_single_max_bytes.saturating_div(2)
+        } else {
+            self.cfg.inline_writes_single_max_bytes
+        };
+
+        let handle = if updates.goodbytes() < inline_threshold {
+            let cfg = self.cfg.clone();
+            let metrics = Arc::clone(&self.metrics);
+            let schemas = schemas.clone();
+
             let span = debug_span!("batch::inline_part", shard = %self.shard_id).or_current();
             mz_ore::task::spawn(
                 || "batch::inline_part",
                 async move {
+                    // Wrap our updates just so the types match.
+                    let updates = BlobTraceUpdates::Row(updates);
```

Comment:
Just to check: we're not worried about a CYA dyncfg for this new bit because we can just turn off the overall structured data one?

Reply:
Exactly, we could add another dyncfg but IMO it's easier to manage just the single flag